src/Nethermind/Nethermind.Merge.AuRa/Shutter/ShutterEonInfo.cs

```diff
@@ -0,0 +1,92 @@
+using Nethermind.Consensus.AuRa.Config;
+using Nethermind.Consensus.Processing;
+using Nethermind.Core.Crypto;
+
+namespace Nethermind.Merge.AuRa.Shutter;
+
+public class ShutterEonInfo
+{
+    public ulong Eon = uint.MaxValue;
+    public Bls.P2 Key;
+    public ulong Threshold;
+    public Address[] Addresses = [];
+    private readonly IReadOnlyBlockTree _readOnlyBlockTree;
+    private readonly ReadOnlyTxProcessingEnvFactory _readOnlyTxProcessingEnvFactory;
+    private readonly IAbiEncoder _abiEncoder;
+    private readonly ILogger _logger;
+    private readonly Address KeyBroadcastContractAddress;
+    private readonly Address KeyperSetManagerContractAddress;
+
+    public ShutterEonInfo(IReadOnlyBlockTree readOnlyBlockTree, ReadOnlyTxProcessingEnvFactory readOnlyTxProcessingEnvFactory, IAbiEncoder abiEncoder, IAuraConfig auraConfig, ILogger logger)
+    {
+        _readOnlyBlockTree = readOnlyBlockTree;
+        _readOnlyTxProcessingEnvFactory = readOnlyTxProcessingEnvFactory;
+        _abiEncoder = abiEncoder;
+        _logger = logger;
+        KeyBroadcastContractAddress = new(auraConfig.ShutterKeyBroadcastContractAddress);
+        KeyperSetManagerContractAddress = new(auraConfig.ShutterKeyperSetManagerContractAddress);
+    }
+
+    public bool Update()
+    {
+        Hash256 stateRoot = _readOnlyBlockTree.Head!.StateRoot!;
+        IReadOnlyTransactionProcessor readOnlyTransactionProcessor = _readOnlyTxProcessingEnvFactory.Create().Build(stateRoot);
+        BlockHeader header = _readOnlyBlockTree.Head!.Header;
+        KeyperSetManagerContract keyperSetManagerContract = new(readOnlyTransactionProcessor, _abiEncoder, KeyperSetManagerContractAddress);
+
+        if (IsSyncing(header))
+        {
+            return false;
+        }
+
+        ulong nextBlockNumber = (ulong)header.Number + 1;
+        ulong eon = keyperSetManagerContract.GetKeyperSetIndexByBlock(header, nextBlockNumber);
+
+        if (Eon != eon)
+        {
+            Eon = eon;
+
+            Address keyperSetContractAddress = keyperSetManagerContract.GetKeyperSetAddress(header, eon);
+            KeyperSetContract keyperSetContract = new(readOnlyTransactionProcessor, _abiEncoder, keyperSetContractAddress);
+
+            if (!keyperSetContract.IsFinalized(header))
+            {
+                if (_logger.IsError) _logger.Error("Cannot use unfinalised keyper set contract.");
+                return false;
+            }
+
+            Threshold = keyperSetContract.GetThreshold(header);
+            Addresses = keyperSetContract.GetMembers(header);
+
+            KeyBroadcastContract keyBroadcastContract = new(readOnlyTransactionProcessor, _abiEncoder, KeyBroadcastContractAddress);
+            byte[] eonKeyBytes = keyBroadcastContract.GetEonKey(_readOnlyBlockTree.Head!.Header, eon);
+            Key = new(eonKeyBytes);
+
+            if (_logger.IsInfo) _logger.Info($"Shutter eon: {Eon} threshold: {Threshold} #keypers: {Addresses.Length}");
```

Comment:
> Changed this so that all fields are updated in one go at the end. Do you think this is sufficient?

It certainly reduces the risk dramatically, but it actually doesn't eliminate it.  [Ref updates are atomic ](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/variables#96-atomicity-of-variable-references) so you can create a class containing all the info and simply update the reference.